contracts/batcher/TwoWayBatcher.sol

```diff
@@ -0,0 +1,99 @@
+//SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.13;
+
+import "@equilibria/root/number/types/UFixed18.sol";
+import "@equilibria/root/token/types/Token18.sol";
+import "@equilibria/root/token/types/Token6.sol";
+import "@equilibria/root/control/unstructured/UReentrancyGuard.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "./Batcher.sol";
+
+contract TwoWayBatcher is UReentrancyGuard, Batcher, ERC20 {
+    /// @dev Event emitted on USDC deposit
+    event Deposit(address indexed account, UFixed18 amount);
+    /// @dev Event emitted on USDC withdraw
+    event Withdraw(address indexed account, UFixed18 amount);
+
+    /// @dev Error thrown on invalid USDC amount
+    error TwoWayBatcherInvalidTokenAmount(UFixed18 amount);
+
+    /**
+     * @notice Initializes the TwoWayBatcher
+     * @dev Called at implementation instantiate and constant for that implementation.
+     * @param reserve EmptySet Reserve Aaddress
+     * @param dsu DSU Token address
+     * @param usdc USDC Token Address
+     */
+    constructor(IEmptySetReserve reserve, Token18 dsu, Token6 usdc)
+    Batcher(reserve, dsu, usdc)
+    ERC20("Batcher Deposit", "BDEP")
+    { } // solhint-disable-line no-empty-blocks
+
+    /**
+     * @notice Deposits USDC for Batcher to use in unwrapping flows
+     * @dev Reverts if `amount` has greater precision than 6 decimals
+     * @param amount Amount of USDC to deposit
+     */
+    function deposit(UFixed18 amount) external nonReentrant {
+        if (!_validToken6Amount(amount)) revert TwoWayBatcherInvalidTokenAmount(amount);
+
+        rebalance();
+
+        USDC.pull(msg.sender, amount, true);
+
+        _mint(msg.sender, UFixed18.unwrap(amount));
+
+        emit Deposit(msg.sender, amount);
+    }
+
+    /**
+     * @notice Withdraws USDC from Batcher
+     * @dev Reverts if `amount` has greater precision than 6 decimals
+     * @param amount Amount of USDC to withdraw
+     */
+    function withdraw(UFixed18 amount) external nonReentrant {
+        if (!_validToken6Amount(amount)) revert TwoWayBatcherInvalidTokenAmount(amount);
+
+        rebalance();
+
+        _burn(msg.sender, UFixed18.unwrap(amount));
+
+        USDC.push(msg.sender, amount);
+
+        emit Withdraw(msg.sender, amount);
+    }
+
+    /**
+     * @notice Rebalances the Batcher to maintain a target balance of USDC and DSU
+     * @dev Maintains a USDC balance of outstanding deposits. Excess USDC is minted as DSU
+     * @param usdcBalance Current Batcher USDC balance
+     */
+    function _rebalance(UFixed18 usdcBalance, UFixed18) override internal {
+        UFixed18 totalSupply = UFixed18.wrap(totalSupply());
+        uint256 balanceToTarget = usdcBalance.compare(totalSupply);
+
+        // totalUSDCLoans == usdcBalance: Do nothing
+        if (balanceToTarget == 1) return;
```

Comment:
technically we don't need this line, but not opposed to keeping it for readability or earlier return